### PR TITLE
CUMULUS-2228: Branching write logic for PDRs

### PR DESCRIPTION
**Pull request by markdboyd**
**Summary:** Summary of changes

Addresses [CUMULUS-2228: Implement branching write logic for PDRs](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2228)

## Changes

* Update branching write logic for sfEventSqsToDbRecords Lambda to add conditional PDR writes to RDS

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Integration tests

### CODE REVIEW

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 1]
```diff
@@ -57,23 +68,49 @@ const hasNoAsyncOpOrExists = async (cumulusMessage, knex) => {
   }, knex, tableNames.asyncOperations);
 };
 
-const hasNoCollectionOrExists = async (cumulusMessage, knex) => {
-  const collectionInfo = getCollectionNameAndVersionFromMessage(cumulusMessage);
-  if (!collectionInfo) {
-    return true;
+const getMessageCollection = async (cumulusMessage, knex) => {
```
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:71]: Executions, PDRs, and granules all need to refer to the collection in RDS (if it exists), so it made sense to me to convert `hasNoCollectionOrExists` into `getMessageCollection`, which actually returns the collection rather than just saying whether it exists.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 2]
```diff
@@ -82,50 +119,124 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
+const savePdr = async ({
+  cumulusMessage,
+  collection,
+  provider,
+  knex,
+  pdrModel = new Pdr(),
+}) => {
+  // If there is no PDR in the message, then there's nothing to do here, which is fine
+  if (!messageHasPdr(cumulusMessage)) {
+    return true;
```
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:162]: I don't like the weird/arbitrary return value here, but the idea was just that the function should return normally in this case

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 3]
```diff
@@ -82,50 +119,124 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
 
+  const results = await Promise.allSettled([
+    executionModel.storeExecutionFromCumulusMessage(cumulusMessage),
+    pdrModel.storePdrFromCumulusMessage(cumulusMessage),
+    granuleModel.storeGranulesFromCumulusMessage(cumulusMessage),
+  ]);
+  const failures = results.filter((result) => result.status === 'rejected');
+  if (failures.length > 0) {
+    const allFailures = failures.map((failure) => failure.reason);
+    log.error(allFailures.join(' '));
```
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:201]: I am not proud of this error handling. I just wanted a way to cleanly log any errors that were encountered. Thoughts?

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 4]
```diff
@@ -82,50 +119,124 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
+      name: getMessagePdrName(cumulusMessage),
+      status: getWorkflowStatus(cumulusMessage),
+      collectionCumulusId: collection.cumulusId,
+      providerCumulusId: provider.cumulusId,
     });
-  } catch (error) {
-    log.error(`Failed to write execution records for ${executionArn}`, error);
-    throw error;
+
+const savePdr = async ({
```
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:153]: I moved to destructuring an options object partially because I think it's cleaner when passing lots of options and also because I wanted pass in the model as an argument, which makes test mocking much easier

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 5]
```diff
@@ -82,50 +119,124 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
+  const isExecutionRDSWriteEnabled = await shouldWriteExecutionToRDS(
+    cumulusMessage,
+    collection,
+    knex
+  );
+
+  // If execution is not written to RDS, then PDRs/granules which reference
+  // execution should not be written to RDS either
+  if (!isExecutionRDSWriteEnabled) {
+    return saveRecordsToDynamoDb(cumulusMessage);
```
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:220]: The logic has been refactored here because I realized: if the execution isn't written to RDS, then neither PDRs or granules should be either, since they both would want to refer to the execution in RDS. Does that seem correct?

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 6]
```diff
@@ -143,16 +254,13 @@ const handler = async (event) => {
...
-    const results = await Promise.allSettled([
-      saveExecution(cumulusMessage, knex),
-      saveGranulesToDb(cumulusMessage),
-      savePdrToDb(cumulusMessage),
-    ]);
-    if (results.some((result) => result.status === 'rejected')) {
+
+    try {
+      await saveRecords(cumulusMessage, knex);
+    } catch (error) {
```
Comment (markdboyd) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:260]: I believe that my implementation preserves the existing behavior where a failure in writing any of the records (to just Dynamo or Dynamo/RDS) results in the message being written to the DLQ. The unit test for this behavior was previously named to suggest that the DLQ was only written to if granule and PDR writes to the database failed, but the actual logic would write when **any failure to write** was encountered:

https://github.com/nasa/cumulus/blob/master/packages/api/tests/lambdas/test-sf-event-sqs-to-db-records.js#L293
https://github.com/nasa/cumulus/blob/master/packages/api/lambdas/sf-event-sqs-to-db-records.js#L57

I kept the unit test for this behavior as previously implemented but updated the name accordingly

File: packages/message/src/PDRs.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+interface PDR {
+  name: string
+  PANSent: boolean
+  PANmessage: string
+}
+
+interface MessageWithPdr extends Message.CumulusMessage {
+  payload: {
+    pdr?: PDR
```
Comment (markdboyd) [inline packages/message/src/PDRs.ts:11]: It isn't entirely clear to me when we want to actually update our core `CumulusMessage` type. I think there was some reason that we intentionally kept `payload` as unknown for that type, because it can be so different in different cases. So it seemed appropriate to create a custom type for this case.

File: packages/message/src/workflows.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import { Message } from '@cumulus/types';
+
+type MessageWithStatus = Message.CumulusMessage & {
```
Comment (markdboyd) [inline packages/message/src/workflows.ts:3]: Same comment about updating core type vs not doing so

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 3 above)
Comment (botanical) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:201]: Are there any other statuses besides `rejected` that indicate a failure?

Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:201]: Not as far as I know: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 2 above)
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:162]: Hmm.   Right now this return looks to be:  

Promise<PromiseResult<AWS.DynamoDB.DocumentClient.UpdateItemOutput, AWS.AWSError>>  || undefined || true 

But our call is an await without any sort of eval - is there a reason not to invert the conditional and just 
```
if(messageHasPdr) {  
  < do the thing >
   return something;
}
```
with/without an explicit return?  

just adding `return` would also resolve the promise as undefined.   I'm not sure what making the return truthy gets us here, unless you're planning on evaluating the 'undefined' response for reasons.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 9]
```diff
@@ -82,50 +119,124 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
   } catch (error) {
     const executionArn = getMessageExecutionArn(cumulusMessage);
-    log.fatal(`Failed to create/update PDR database record for execution ${executionArn}: ${error.message}`);
+    log.fatal(`Failed to create/update granule records for execution ${executionArn}: ${error.message}`);
     throw error;
   }
 };
 
-const saveGranulesToDb = async (cumulusMessage) => {
+const saveRecordsToDynamoDb = async (cumulusMessage) => {
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:188]: Noting for my own sanity - this is a re-implementation of the Dynamo logic originally present if isExecutionRDSWriteEnabled is set, instead of implementing something that makes the various 'save' calls conditional.   

This approach seems fine.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 3 above)
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:201]: That's a reasonable change to this, presuming that all of the models reject on failures, which they appear to.    I'm not aware of anything cleaner short of a function abstraction that would accomplish this.  

It's worth noting (without suggesting action)  though, in the case of a granule failure, I believe you're only going to get the *first* failed granule from storeGranulesFromCumulusMessage as the model is returning 

```
    return Promise.all(granuleRecords.map(this._validateAndStoreGranuleRecord, this));
```

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 10]
```diff
@@ -82,50 +119,124 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
+  const results = await Promise.allSettled([
+    executionModel.storeExecutionFromCumulusMessage(cumulusMessage),
+    pdrModel.storePdrFromCumulusMessage(cumulusMessage),
+    granuleModel.storeGranulesFromCumulusMessage(cumulusMessage),
+  ]);
+  const failures = results.filter((result) => result.status === 'rejected');
+  if (failures.length > 0) {
+    const allFailures = failures.map((failure) => failure.reason);
+    log.error(allFailures.join(' '));
+    throw new Error('Failed writing some records to Dynamo');
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:202]: We could probably make this a bit more descriptive though, given we have `allFailures`

Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:202]: Also what does the record in the DLQ look like for this failure as compared to the implementation on main? 

*edit - answered below*

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 5 above)
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:220]: Yes... but is that true for granules as well?    We can save that for the next PR.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 6 above)
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:260]: Good catch - I believe that was indeed the original intent.

File: packages/api/models/executions.js [hunk 11]
```diff
@@ -40,7 +43,7 @@ class Execution extends Manager {
     const arn = getMessageExecutionArn(cumulusMessage);
     if (isNil(arn)) throw new Error('Unable to determine execution ARN from Cumulus message');
 
-    const status = get(cumulusMessage, 'meta.status');
+    const status = getWorkflowStatus(cumulusMessage);
```
Comment (Jkovarik) [inline packages/api/models/executions.js:46]: I'm a bit torn on these abstractions.   On the one hand, not using strings everywhere is a good abstraction, on the other hand, `workflowStatus` is an indirect reference to meta.status, but changing it to getMetaStatus defeats the abstraction....and also we're putting things in 'meta' like 'status' that users could easily accidently overwrite.  ( https://nasa.github.io/cumulus/docs/workflows/cumulus-task-message-flow#docsNav) 

🤔   

This is a sidebar to this PR, as given while I type this I'm realizing it's more about the `meta` namespace than the abstraction itself I don't like.   Leaving here for posterity/to trigger a discussion.

**Review by Jkovarik — COMMENTED** [review]
This looks realy good - I want to do some manual tests/unit test review tomorrow, but no major blocking comments in the main code.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 2 above)
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:162]: Just using `return` causes an `inconsistent-return` error from ESlint. As does inverting the logic. I'm tempted to ignore ESlint as being capricious, except that I feel like it often surfaces poorly written code. So I'm not sure what to do here

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 3 above)
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:201]: Yeah, we're going to have to refactor the granule write logic which I noticed. See the discussion in the cumulus-ix channel about a transaction for writing all granules vs a transaction per granule. But I'm going to leave that for the next PR

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 5 above)
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:220]: I'm fairly certain it would also be true for granules? Granules refer to an execution, so if the execution wasn't written to RDS, those granule RDS writes would fail so we shouldn't attempt any RDS write for them. Does that sound right?

File: packages/api/models/executions.js (hunk 11 above)
Comment (markdboyd) [reply packages/api/models/executions.js:46]: OK. I guess I don't totally follow. The only point of `getWorkflowStatus` and other functions in the `message` package is to have trivial helpers for extracting information from the workflow message, which we're now doing in at least two places (Dynamo models and RDS write code) and possibly more in the future. Is the concern the function name?

File: packages/message/src/PDRs.ts (hunk 7 above)
Comment (markdboyd) [reply packages/message/src/PDRs.ts:11]: OK re: sticking with an override for this case.

How about`MessageWithOptionalPayloadPdr` for the name?

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 2 above)
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:162]: There's probably a legit case for `treatUndefinedAsUnspecified` being true, but I'm curious how the downstream code that's returning undefined is passing lint then.    🤔    Anyway, `null` is explicit, is that the right pattern for this sort of return, given that lint rule? 🤔

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 5 above)
Comment (Jkovarik) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:220]: Probably.  The situation I'm still worried about is a situation where we notify CMR about a granule, but fail to have a granule record in Core.     From a data model/db perspective, obviously you want a complete record, but I have some disquiet about what users are going to want in this regard.

  Right now in Dynamo if an execution write fails the granule *may or may not*  write( thanks promise.all), so that behavior is suboptimal anyway. 

That concern is *probably* best relegated to the "what do we do about DLQ messages" and/or "reconciliation report handling" feature/discussion/ether though.

File: packages/api/models/executions.js (hunk 11 above)
Comment (Jkovarik) [reply packages/api/models/executions.js:46]: Yeah, sorry, I could have expressed that better.      Reading through it with review blinders on, I'm not sure I  know what 'workflowStatus' is as an entity on a Cumulus message, except via the helper source that tells me it means meta.status.   

(having it in `meta` is in userspace is a separate discussion I'm sure we already had, but `status` could easily be accidently overrode by a user as we don't define anything in meta as reserved) 

tldr:  My concern would be resolved if this sort of method was named getMetaStatus, but following *that* pattern seems like a thin reason to wrap cumulusMessage?.meta?.status in TS as an update, although perhaps justified in vanilla node.    

It'd also be resolved if it were in cumulus_meta and that object were defined better-*er*, but that starts to *really* feel outside scope.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 5 above)
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js:220]: The difference is that in Dynamo, we're not making the granule write dependent on the execution write. The granule dynamo write may still fail for _reasons_ though.

But yeah, agree that this is really a question of how to handle/surface errors produced by this Lambda, which I think is outside this ticket

File: packages/message/src/PDRs.ts (hunk 7 above)
Comment (Jkovarik) [reply packages/message/src/PDRs.ts:11]: Sure, that seems fine.

File: packages/api/models/executions.js (hunk 11 above)
Comment (markdboyd) [reply packages/api/models/executions.js:46]: I think the reason to wrap `cumulusMessage?.meta?.status` in a helper is just to avoid having that logic duplicated, decreasing the chances (however small) of bugs. Let me know if you disagree though. I'm fine with renaming it to `getMetaStatus`.

Comment (Jkovarik) [reply packages/api/models/executions.js:46]: And to be clear, this is absolutely nit territory.    We should time box more thought on this to, say, 15 minutes.

Comment (Jkovarik) [reply packages/api/models/executions.js:46]: > I'm fine with renaming it to getMetaStatus.

That seems fine.  If it causes annoyance and we actually ever move it from meta, we can put a PR up and discuss there.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 3 above)
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js]: I realize we were comfortable with the error handling, but I didn't like the awkward merging of error messages, so I updated this code to throw an AggregateError instead:

093da4197c7569124a851f2eba866505498c51b7

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 2 above)
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js]: I realized that we're going to want to explicitly return the PDR's cumulusId so that a granule can refer to it (implementation of that write logic in the next PDR), so this PDR write logic needed a bit of a refactor. Given the refactor, I think returning `undefined` when there is no PDR on the message is appropriate, so that the function either returns `undefined`, returns the PDR cumulusId, or throws an error. Also, using `undefined` over `null` because we have ESlint warnings on using `null`

29fb20d142f112e268888b6005cc4be50e484a4e

File: packages/message/src/PDRs.ts (hunk 7 above)
Comment (markdboyd) [reply packages/message/src/PDRs.ts:11]: bdce8ab59a5a77fec0e36b405ac0caaee97a7120

File: packages/api/lambdas/sf-event-sqs-to-db-records.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (markdboyd) [reply packages/api/lambdas/sf-event-sqs-to-db-records.js]: e192e56de314ba0d66c428ecede395f85d02e03a

File: packages/api/models/executions.js (hunk 11 above)
Comment (markdboyd) [reply packages/api/models/executions.js]: 0f84f40d83d9a2d158d4933177eb43d92b64a113

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 12]
```diff
@@ -82,52 +120,132 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
+  // execution should not be written to RDS either
+  if (!isExecutionRDSWriteEnabled) {
+    return writeRecordsToDynamoDb(cumulusMessage);
+  }
+
+  const provider = await getMessageProvider(cumulusMessage, knex);
+
   try {
-    await granuleModel.storeGranulesFromCumulusMessage(cumulusMessage);
+    console.log('about to write execution');
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:232]: I realize this is in-progress work/might be useful debugging, but we should remove these before 2228/make them logger statements before the work is complete.

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 13]
```diff
@@ -145,16 +263,15 @@ const handler = async (event) => {
...
     const executionEvent = parseSQSMessageBody(message);
     const cumulusMessage = await getCumulusMessageFromExecutionEvent(executionEvent);
-    const results = await Promise.allSettled([
-      saveExecution(cumulusMessage, knex),
-      saveGranulesToDb(cumulusMessage),
-      savePdrToDb(cumulusMessage),
-    ]);
-    if (results.some((result) => result.status === 'rejected')) {
+
+    console.log(message, 'message');
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js]: Same comment as above re: logging

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 14]
```diff
@@ -82,52 +120,132 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
...
 
+  const results = await Promise.allSettled([
+    executionModel.storeExecutionFromCumulusMessage(cumulusMessage),
+    pdrModel.storePdrFromCumulusMessage(cumulusMessage),
+    granuleModel.storeGranulesFromCumulusMessage(cumulusMessage),
+  ]);
+  const failures = results.filter((result) => result.status === 'rejected');
+  if (failures.length > 0) {
+    const allFailures = failures.map((failure) => failure.reason);
+    const aggregateError = new AggregateError(allFailures);
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:205]: This is interesting (the good kind)

File: packages/api/tests/lambdas/test-sf-event-sqs-to-db-records.js [hunk 15]
```diff
@@ -287,194 +314,162 @@ test.serial('hasNoAsyncOpOrExists returns false if async operation does not exis
...
-      cumulus_meta: {
-        cumulus_version: '3.0.0',
-        parentExecutionArn,
-      },
-    }, knex)
+    await shouldWriteExecutionToRDS(cumulusMessage, { cumulusId: collectionCumulusId }, knex)
+  );
+});
+
+test('shouldWriteExecutionToRDS returns false if collection record is not defined', async (t) => {
```
Comment (Jkovarik) [inline packages/api/tests/lambdas/test-sf-event-sqs-to-db-records.js:434]: Hmm, this makes me wonder if isPostRDSDeploymentExecution should be throwing at all in this case. 

Is there a reason we shouldn't raise a warning message not throw in that method, so that another throw elsewhere doesn't cause the lambda to halt?

File: packages/api/lambdas/sf-event-sqs-to-db-records.js [hunk 16]
```diff
@@ -57,23 +69,49 @@ const hasNoAsyncOpOrExists = async (cumulusMessage, knex) => {
...
+) => {
   try {
-    const executionIsPostDeployment = isPostRDSDeploymentExecution(cumulusMessage);
-    if (!executionIsPostDeployment) return executionIsPostDeployment;
+    const isExecutionPostDeployment = isPostRDSDeploymentExecution(cumulusMessage);
+    if (!isExecutionPostDeployment) return false;
+    if (!isRecordDefined(collection)) return false;
 
     const results = await Promise.all([
       hasNoParentExecutionOrExists(cumulusMessage, knex),
```
Comment (Jkovarik) [inline packages/api/lambdas/sf-event-sqs-to-db-records.js:113]: Not this PR so not blocking, but should these throw such that this doesn't await the other promise if there's a failure? 

That might also make checking if they're all true cleaner.

**Review by Jkovarik — APPROVED** [review]
Approving - just a couple of comments for consideration, and:

1) I can't recall if we've had this conversation or not, but we should probably update these units to create/provision a db where it's actually writing to avoid cleanup/persistence issues in the unit tests.  
2) Again not *this* PR, but this 'main' .js is getting pretty lengthy.  We should consider splitting it up a bit.
